demos/connector/neo4j/load-cora-into-neo4j.ipynb

```diff
@@ -0,0 +1,413 @@
+   "source": [
+    "# gather all features into lists under 'features' column.\n",
+    "node_list[\"features\"] = node_list[feature_names].values.tolist()\n",
+    "\n",
+    "node_list = node_list.drop(columns=feature_names)\n",
+    "node_list[\"id\"] = node_list.index\n",
+    "node_list.head(5)"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "## Ingest data into Neo4J database\n",
+    "\n",
+    "We define the graph schema as below:\n",
+    "\n",
+    "- Each vertex represents a paper\n",
+    "    + subject (String): the class where each subject belongs to. There are seven classes in total. \n",
+    "    + features (List[int]): 1D-vector represents the presence of each words in the dictionary.\n",
+    "    + ID (int): id of each paper. (**Note**: this ID attribute is different from the Neo4J id, i.e., the id of each node or relationship which Neo4J automatically assigns with). \n",
+    "    \n",
+    "- Each *directed* edge represents a citation. Each edge points to the paper being cited.\n",
+    "\n",
+    "As the Cora dataset is small, we could use Cypher queries and execute the transactions via a Python-supported driver.\n",
+    "\n",
+    "For bigger dataset, this loading job might take very long, so it is more convenient to use ```neo4j-admin import ``` tool, [tutorial here](https://neo4j.com/docs/operations-manual/current/tutorial/import-tool/)."
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 7,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "import py2neo\n",
+    "\n",
+    "# Create the Neo4J Graph database object\n",
+    "graph = py2neo.Graph(\"http://localhost:7474\", username=\"neo4j\", password=\"cora\")"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "Delete the existing edges and relationships in the current database."
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 8,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "empty_db_query = \"\"\"\n",
+    "    MATCH(n) DETACH\n",
+    "    DELETE(n)\n",
+    "    \"\"\"\n",
+    "\n",
+    "tx = graph.begin(autocommit=True)\n",
+    "tx.evaluate(empty_db_query)"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "Load all nodes to the graph database."
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 9,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "loading_node_query = \"\"\"\n",
+    "    UNWIND $node_list as node\n",
+    "    CREATE( e: paper {\n",
+    "        ID: toInteger(node.id),\n",
+    "        subject: node.subject,\n",
+    "        features: node.features\n",
+    "    })\n",
+    "    \"\"\"\n",
+    "\n",
+    "# For efficient loading, we will load batch of nodes into Neo4J.\n",
+    "batch_len = 500\n",
+    "\n",
+    "for batch_start in range(0, len(node_list), batch_len):\n",
+    "    batch_end = min(len(node_list), batch_start + batch_len)\n",
```

Comment:
Python slicing is relaxed so this `min` work isn't necessary: the endpoint is automatically clipped at the maximum value, e.g.

```
>>> pd.DataFrame(["a", "b", "c", "d"], columns=["X"]).iloc[2:99999999]
   X
2  c
3  d
```

That is, we can remove this line and write:

```python
records = node_list.iloc[batch_start:batch_start + batch_len].to_dict("records")
```

or, just simplify this line

```python
batch_end = batch_start + batch_len
records = node_list.iloc[batch_start:batch_end].to_dict("records")
```

The same applies to the edge loading.